packages/fxa-payments-server/src/components/CouponForm/index.tsx

```diff
@@ -107,8 +109,29 @@ export const CouponForm = ({
   );
   const [error, setError] = useState<CouponErrorMessageType | null>(null);
   const [checkingCoupon, setCheckingCoupon] = useState(false);
+  const { queryParams } = useContext(AppContext);
 
   const onFormMounted = useCallback(() => couponMounted({ planId }), [planId]);
+
+  // check if coupon code was included in URL
+  // if true and valid, apply coupon code on page load
+  useEffect(() => {
+    (async () => {
```

Comment:
Q: Does this need to be `async` (since I don't see any promise chaining or `await`ing)?
If not, do we need the inner IIFE, or can we just migrate all the code into the `useEffect()` and delete a closure?

Reply:
This was done because of L123 `const coupon = await checkPromotionCode(planId, queryParams.coupon);` where it has to wait for coupon validation before proceeding. I tried to find a way to simplify it and not repeat code, as the try/catch is nearly identical to L151-L161, save for the coupon code and setting an error message, but couldn't find a way around it that would fit within the scope of the ticket. 😕 Let me know if you have any ideas. Would love to optimize it if possible! 🙂

Reply:
Oy. Not sure how I missed the `await` on L123, but yeah, IIFE might be best since it doesn't look like you can easily use `async` `useEffect()` hooks, per:
- https://devtrium.com/posts/async-functions-useeffect
- https://ultimatecourses.com/blog/using-async-await-inside-react-use-effect-hook